### PR TITLE
[VPU][TESTS][GNA] Fix dynamic models import on VPU

### PR DESCRIPTION
**Pull request by AndrewBakalinIntel**
Ticket - #-43979
Changes:
- Fix for importing of dynamic models (remove internal shape inputs/outputs insertion to containers that are visible for users & fill `descFromPlugin` field)
- Extract import/export tests from GNA to shared part in order to reuse it for other plugins

### CODE REVIEW

**Comment by AndrewBakalinIntel** [comment]
@andrejsokolov @gladilov-gleb could you please take a look?
@iefode could you please take a look at tests part?

File: inference-engine/tests/functional/plugin/shared/include/import_export_tests/import_export_base.hpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+typedef std::tuple<
+    InferenceEngine::Precision,         // Network Precision
+    std::string,                        // Target Device
+    std::map<std::string, std::string>, // Export Configuration
+    std::map<std::string, std::string>  // Import Configuration
+> exportImportNetworkParams;
+
+namespace LayerTestsDefinitions {
+
+class ImportNetworkTestBase : public testing::WithParamInterface<exportImportNetworkParams>,
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/shared/include/import_export_tests/import_export_base.hpp:20]: Looks like, we should move this class to the `funcTestUtils`

**Review by iefode — CHANGES_REQUESTED** [review]

File: inference-engine/src/vpu/graph_transformer/src/blob_reader.cpp [hunk 2]
```diff
@@ -95,8 +95,11 @@ void BlobReader::parse(const std::vector<char>& blob) {
         ie::InputInfo input;
         input.setInputData(std::make_shared<ie::Data>(inputData));
 
-        _networkInputs[input.name()]    = std::make_shared<ie::InputInfo>(input);
+        if (input.name().find("@shape") == std::string::npos) {
```
Comment (gladilov-gleb) [inline inference-engine/src/vpu/graph_transformer/src/blob_reader.cpp:98]: Nitpick: probably it worthies to come up with a dedicated method or something. Hardcoding logic (that name contains @shape) does not look too good

File: inference-engine/src/vpu/graph_transformer/src/blob_reader.cpp [hunk 3]
```diff
@@ -142,8 +145,11 @@ void BlobReader::parse(const std::vector<char>& blob) {
         ie::TensorDesc ieDesc = DataDesc(dataType, dimsOrder, vpuDims).toTensorDesc();
         ie::Data outputData(outputName, ieDesc);
 
-        _networkOutputs[outputData.getName()]    = std::make_shared<ie::Data>(outputData);
+        if (outputData.getName().find("@shape") == std::string::npos) {
```
Comment (gladilov-gleb) [inline inference-engine/src/vpu/graph_transformer/src/blob_reader.cpp:148]: Looks like a copy-paste from inputs. Should we remove code duplication?

File: inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    if (inputStream.fail()) {
+        FAIL() << "Cannot open file to import model: exported_model.blob";
+    }
+    auto importedNetwork = core->ImportNetwork(inputStream, targetDevice, configuration);
+    for (const auto& next_input : importedNetwork.GetInputsInfo()) {
+        ASSERT_NO_THROW(executableNetwork.GetInputsInfo()[next_input.first]);
+    }
+    for (const auto& next_output : importedNetwork.GetOutputsInfo()) {
+        ASSERT_NO_THROW(executableNetwork.GetOutputsInfo()[next_output.first]);
+    }
```
Comment (gladilov-gleb) [inline inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp:55]: I see here a check that all inputs/outputs of imported model are present in executable network. I expect that check be more strict: inputs/outputs should be equal

File: inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    auto importedOutputs = CalculateImportedNetwork(importedNetwork);
+    Compare(importedOutputs, actualOutputs);
+}
+
+std::vector<std::vector<std::uint8_t>> ImportNetworkTestBase::CalculateImportedNetwork(InferenceEngine::ExecutableNetwork& importedNetwork) {
```
Comment (gladilov-gleb) [inline inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp:60]: What is the reason for such function? Should there be a component for inferring agnostic to the fact that model has been imported or loaded?

File: inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    executableNetwork.Export("exported_model.blob");
+
+    const auto& actualOutputs = GetOutputs();
+    auto referenceOutputs = CalculateRefs();
+    Compare(referenceOutputs, actualOutputs);
+
+    for (auto const& configItem : importConfiguration) {
+        configuration[configItem.first] = configItem.second;
+    }
+    std::fstream inputStream("exported_model.blob", std::ios_base::in | std::ios_base::binary);
```
Comment (gladilov-gleb) [inline inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp:45]: So after running tests artifacts are present on the system? Should we remove this file once test has been successfully executed?

File: inference-engine/tests/functional/plugin/shared/include/import_export_tests/import_export_base.hpp (hunk 1 above)
Comment (AndrewBakalinIntel) [reply inference-engine/tests/functional/plugin/shared/include/import_export_tests/import_export_base.hpp:20]: Thanks! Done.

File: inference-engine/src/vpu/graph_transformer/src/blob_reader.cpp (hunk 2 above)
Comment (AndrewBakalinIntel) [reply inference-engine/src/vpu/graph_transformer/src/blob_reader.cpp]: I'm not sure I've made some final solution on this, but at least I've moved this hardcoded `@shape` to a dedicated place. Please take a look.

File: inference-engine/src/vpu/graph_transformer/src/blob_reader.cpp (hunk 3 above)
Comment (AndrewBakalinIntel) [reply inference-engine/src/vpu/graph_transformer/src/blob_reader.cpp]: Agreed, done.

File: inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp (hunk 4 above)
Comment (AndrewBakalinIntel) [reply inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp]: What do you mean `should be equal`? The underlying objects are still different and BTW they don't have `operator==`

File: inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp (hunk 5 above)
Comment (AndrewBakalinIntel) [reply inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp]: Actually the existing code may be reused here. Fixed.

File: inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp (hunk 6 above)
Comment (AndrewBakalinIntel) [reply inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp]: I've replaced `fstream` with `stringstream` (as it's done for other tests) so now there are now artifacts creating.

File: inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp (hunk 4 above)
Comment (gladilov-gleb) [reply inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp]: I mean that any input or output presented in loaded model presents in imported as well and vice versa. Currently check is done only in one direction: all inputs and output in imported model presents in loaded.

Comment (gladilov-gleb) [reply inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp]: Nitpick:

> they don't have operator==

Do you think it is possible to provide it?

File: inference-engine/tests/functional/plugin/myriad/shared_tests_instances/import_export_tests/import_nonzero.cpp [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+using namespace LayerTestsDefinitions;
+
+namespace {
+
+TEST_P(ImportNonZero, CompareWithRefImpl) {
+    Run();
+};
+
+const std::vector<InferenceEngine::Precision> netPrecisions = {
+        InferenceEngine::Precision::FP32,
```
Comment (andrejsokolov) [inline inference-engine/tests/functional/plugin/myriad/shared_tests_instances/import_export_tests/import_nonzero.cpp:17]: There should be only these precisions to test?

Comment (AndrewBakalinIntel) [reply inference-engine/tests/functional/plugin/myriad/shared_tests_instances/import_export_tests/import_nonzero.cpp:17]: Actually this test doesn't depend on precision at all, the network should only have dynamic output in order to reproduce the initial problem. So I've finally decided to leave only one precision to remove unnecessary test cases.

File: inference-engine/src/vpu/graph_transformer/src/blob_reader.cpp (hunk 3 above)
Comment (AndrewBakalinIntel) [reply inference-engine/src/vpu/graph_transformer/src/blob_reader.cpp]: I've fixed the previous version of this code to get rid of `bool isInput`, so now it looks better. Please take a look once more.

File: inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp (hunk 4 above)
Comment (AndrewBakalinIntel) [reply inference-engine/tests/functional/plugin/shared/src/import_export_tests/import_export_base.cpp]: I've added another checks that the number of inputs/outputs are equal, this makes one directional check enough for this. 
From my perspective these objects are too complicated to provide `operator==` for them.

**Review by andrejsokolov — APPROVED** [review]

**Review by gladilov-gleb — APPROVED** [review]

**Comment by AndrewBakalinIntel** [comment]
@iefode could you please take a look once more?

File: inference-engine/tests/functional/plugin/shared/include/import_export_tests/import_export_base.hpp (hunk 1 above)
Comment (iefode) [reply inference-engine/tests/functional/plugin/shared/include/import_export_tests/import_export_base.hpp:20]: Again, I said about the `funcTestUtils`

**Review by iefode — APPROVED** [review]
In general, LGTM, beside one comment

**Comment by iefode** [comment]
Please, fix the CI and i will merge it

**Comment by AndrewBakalinIntel** [comment]
@iefode CI is green.

**Pull request merged**